### PR TITLE
data versioning phase 1

### PR DESCRIPTION
**Pull request by ravishetye**
{create, get, delete} X {dataset, dataset version}

there is GetExperimentRunsByDatasetVersionId but I think it can be supported by FindExperimentRuns and the extra call might not be required.

### CODE REVIEW

File: protos/public/modeldb/Dataset.proto [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    string name = 2;
+    string owner = 3;
+    string description = 4;
+    repeated string tags = 5; 
+    bool publicly_visible = 6;
+    DatasetType dataset_type = 7;
+    repeated KeyValue attributes = 8;
+}
+
+enum DatasetType{
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: 👀

File: protos/public/modeldb/Dataset.proto [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+message RawDatasetInfo {
+    uint64 size = 1;
+    repeated string  features = 2;
+    uint64 num_records = 3;
+    string object_path = 4;
+    string check_sum = 5;
+}
+
+enum PathLocationType{
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: 👀

File: protos/public/modeldb/CommonService.proto [hunk 3]
```diff
@@ -38,7 +38,6 @@ message ArtifactTypeEnum{
         DATA = 3;
         BLOB = 4;
         STRING = 5;
-        CODE = 6;
```
Comment (convoliution) [inline protos/public/modeldb/CommonService.proto]: Wait make sure that's not deleted!

File: protos/public/modeldb/CommonService.proto [hunk 4]
```diff
@@ -47,6 +46,7 @@ message Artifact {
     string path = 2;
     bool path_only = 3;
     ArtifactTypeEnum.ArtifactType artifact_type = 4;
+    string dataset_version_id = 5;
```
Comment (mvartakAtVerta) [inline protos/public/modeldb/CommonService.proto]: call this linked_obj_id or something similar?

File: protos/public/modeldb/Dataset.proto (hunk 1 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: what is the comment here?

File: protos/public/modeldb/CommonService.proto (hunk 4 above)
Comment (ravishetye) [reply protos/public/modeldb/CommonService.proto]: linked_artifact_id

File: protos/public/modeldb/CommonService.proto (hunk 3 above)
Comment (ravishetye) [reply protos/public/modeldb/CommonService.proto]: added it in

File: protos/public/modeldb/Dataset.proto (hunk 2 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: nested

File: protos/public/modeldb/Dataset.proto (hunk 1 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: nested

File: protos/public/modeldb/ExperimentRunService.proto [hunk 5]
```diff
@@ -836,4 +849,10 @@ service ExperimentRunService {
             get: "/v1/experiment-run/setParentExperimentRunId"
         };
     };
+
+    rpc getExperimentRunsByDatasetVersionId (GetExperimentRunsByDatasetVersionId) returns (GetExperimentRunsByDatasetVersionId.Response){
```
Comment (ad-47) [inline protos/public/modeldb/ExperimentRunService.proto]: this service not implemented in this (https://github.com/VertaAI/verta-backend/blob/rs/dataset-as-fcc/src/main/java/com/mitdbg/modeldb/experimentRun/ExperimentRunServiceImpl.java) file.

File: protos/public/modeldb/Dataset.proto [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    string object_path = 4;
+    string check_sum = 5;
+}
+
+message PathLocationTypeEnum{
+    enum PathLocationType{
+        LOCAL_FILE_SYSTEM = 0;// default
+        NETWORK_FILE_SYSTEM = 1;
+        HADOOP_FILE_SYSTEM = 2;
+        s3_FILE_SYSTEM = 3;
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: Is there a reason this uses a lowercase 's'? Amazon's branding uses a capital 'S' anyway.

File: protos/public/modeldb/Dataset.proto [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+//getAllDataset
+message GetAllDatasets {
+    //For pagination
+    int32 page_number = 6;
+    int32 page_limit = 7;
+    bool ascending = 9;
+    string sort_key = 10;
+    message Response {
+        repeated Dataset datasets = 1;
+        int64 total_records = 2;
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: Should this be `uint64`?
(see also line 166, and line 536 in ExperimentRunService)

File: protos/public/modeldb/Dataset.proto [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    string name = 2;
+    string owner = 3;
+    string description = 4;
+    repeated string tags = 5; 
+    DatasetVisibilityEnum.DatasetVisibility dataset_visibility = 6;
+    DatasetTypeEnum.DatasetType dataset_type = 7;
+    repeated KeyValue attributes = 8;
+}
+
+message DatasetTypeEnum{
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: Missing space before opening bracket.
(see also lines 23, 30, 79, 88, 188, 195)

File: protos/public/modeldb/Dataset.proto [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    string dataset_id = 3;
+    uint64 time_logged = 4;
+    string description = 5;
+    repeated string tags = 6;
+    DatasetVisibilityEnum.DatasetVisibility dataset_version_visibility = 7;
+    DatasetTypeEnum.DatasetType dataset_type = 8;// this acts as a quick check on which type of DataSetInfo to look at for more details.
+    repeated KeyValue attributes = 9;
+    string owner = 10;
+    uint64 version = 11;
+    oneof  dataset_version_info {
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: Extra space after `oneof`.
(see also lines 73, 146)

File: protos/public/modeldb/Dataset.proto [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    }
+}
+
+message QueryDatasetInfo {
+    string query = 1;
+    string query_template = 2;
+    repeated QueryParameter query_parameters = 3;
+    string data_source_uri = 4;
+    uint64 execution_timestamp = 5;
+    uint64 records = 6;
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: I don't have context for this, but should it be either `num_records` (like line 74) or `total_records` (like line 124)?

File: protos/public/modeldb/Dataset.proto [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    ValueTypeEnum.ValueType parameter_type = 2;
+    google.protobuf.Value value = 3;
+}
+
+message RawDatasetInfo {
+    uint64 size = 1;
+    repeated string  features = 2;
+    uint64 num_records = 3;
+    string object_path = 4;
+    string check_sum = 5;
```
Comment (convoliution) [inline protos/public/modeldb/Dataset.proto]: Should be one word `checksum` (like line 98).

File: protos/public/modeldb/Dataset.proto (hunk 6 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: fixed.

File: protos/public/modeldb/Dataset.proto (hunk 7 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: fixed.

File: protos/public/modeldb/Dataset.proto (hunk 8 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: fixed

File: protos/public/modeldb/Dataset.proto (hunk 9 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: fixed

File: protos/public/modeldb/Dataset.proto (hunk 10 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: fixed.

File: protos/public/modeldb/Dataset.proto (hunk 11 above)
Comment (ravishetye) [reply protos/public/modeldb/Dataset.proto]: fixed

File: protos/public/modeldb/ExperimentRunService.proto (hunk 5 above)
Comment (ravishetye) [reply protos/public/modeldb/ExperimentRunService.proto]: yes this will be implemented later.

**Pull request merged**